user-ops-indexer/user-ops-indexer-logic/src/indexer/rpc_utils.rs

```diff
@@ -0,0 +1,158 @@
+                            tracer_config: None,
+                            timeout: Some("60s".to_string()),
+                        },
+                    )
+                    .await?;
+
+                match geth_trace {
+                    GethTrace::Known(GethTraceFrame::CallTracer(root)) => {
+                        Ok(flatten_geth_trace(root))
+                    }
+                    _ => Err(ProviderError::CustomError(
+                        "can't parse geth trace result".to_string(),
+                    )),
+                }
+            }
+            _ => {
+                let traces = self
+                    .trace_transaction(tx_hash)
+                    .await?
+                    .into_iter()
+                    .filter_map(|t| match t.action {
+                        Action::Call(call) => Some(CommonCallTrace {
+                            typ: match call.call_type {
+                                CallType::Call => TraceType::Call,
+                                CallType::CallCode => TraceType::CallCode,
+                                CallType::DelegateCall => TraceType::DelegateCall,
+                                CallType::StaticCall => TraceType::StaticCall,
+                                CallType::None => TraceType::Other,
+                            },
+                            from: call.from,
+                            to: Some(call.to),
+                            input: call.input,
+                        }),
+                        Action::Create(create) => Some(CommonCallTrace {
+                            typ: TraceType::Create,
+                            from: create.from,
+                            to: None,
+                            input: create.init,
+                        }),
+                        _ => None,
+                    })
+                    .collect();
+
+                Ok(traces)
+            }
+        }
+    }
+}
+
+fn flatten_geth_trace(root: CallFrame) -> Vec<CommonCallTrace> {
+    let mut path = Vec::from([(&root, 0)]);
+    let mut res = Vec::new();
+
+    while let Some((frame, idx)) = path.pop() {
+        if idx == 0 {
+            res.push(CommonCallTrace {
+                typ: match frame.typ.as_str() {
+                    "CALL" => TraceType::Call,
+                    "CALLCODE" => TraceType::CallCode,
+                    "STATICCALL" => TraceType::StaticCall,
+                    "DELEGATECALL" => TraceType::DelegateCall,
+                    "CREATE" => TraceType::Create,
+                    _ => TraceType::Other,
+                },
+                from: frame.from,
+                to: frame.to.clone().and_then(|to| to.as_address().cloned()),
+                input: frame.input.clone(),
+            });
+        }
+        if let Some(calls) = &frame.calls {
+            if calls.len() > idx {
+                path.push((&frame, idx + 1));
+                path.push((&calls[idx], 0));
+            }
+        }
```

Comment:
Maybe I'm missing something, but why to make such complex schema for adding all subcalls instead of something like this:
```suggestion
        if let Some(calls) = frame.calls {
            path.extend(calls)
        }
        res.push(CommonCallTrace {
            typ: match frame.typ.as_str() {
                "CALL" => TraceType::Call,
                "CALLCODE" => TraceType::CallCode,
                "STATICCALL" => TraceType::StaticCall,
                "DELEGATECALL" => TraceType::DelegateCall,
                "CREATE" => TraceType::Create,
                _ => TraceType::Other,
            },
            from: frame.from,
            to: frame.to.and_then(|to| to.as_address().cloned()),
            input: frame.input,
        });
```